pkg/didcomm/crypto/legacy/authcrypt/encrypt.go

```diff
@@ -0,0 +1,219 @@
+func (c *Crypter) buildJSON(protectedBytes, nonce, cipherText, tag []byte) ([]byte, error) {
+
+	envelope := &Envelope{
+		Protected:  base64.URLEncoding.EncodeToString(protectedBytes),
+		IV:         base64.URLEncoding.EncodeToString(nonce),
+		CipherText: base64.URLEncoding.EncodeToString(cipherText),
+		Tag:        base64.URLEncoding.EncodeToString(tag),
+	}
+
+	jsonOut, err := json.Marshal(envelope)
+	if err != nil {
+		return nil, err
+	}
+
+	return jsonOut, nil
+}
+
+func (c *Crypter) buildProtected(recipients []Recipient) ([]byte, error) {
+
+	var enc string
+	switch c.alg {
+	case C20P:
+		enc = "chacha20poly1305_ietf"
+	case XC20P:
+		enc = "xchacha20poly1305_ietf"
+	}
+
+	protected := Protected{
+		Enc:        enc,
+		Typ:        "JWM/1.0",
+		Alg:        "Authcrypt",
+		Recipients: recipients,
+	}
+
+	protectedBytes, err := json.Marshal(protected)
+	if err != nil {
+		return nil, err
+	}
+
+	m, err := prettyPrint(protectedBytes)
+	if err != nil {
+		return nil, err
+	}
+	print(m, "\n")
+
+	return protectedBytes, nil
+}
+
+func (c *Crypter) buildRecipients(cek *[chacha20poly1305.KeySize]byte) ([]Recipient, error) {
+	var encodedRecipients []Recipient
+
+	for _, recKey := range c.recipients {
+		recipient, err := c.buildRecipient(cek, recKey)
+		if err != nil {
+			return nil, err
+		}
+
+		encodedRecipients = append(encodedRecipients, *recipient)
+	}
+
+	return encodedRecipients, nil
+}
+
+// Note: the spec is incorrect with the parameters for encrypted_key: the CEK is the message to encrypt
+// https://github.com/hyperledger/aries-cloudagent-python/blob/master/aries_cloudagent/wallet/crypto.py#L277
+
+/*
+{
+	"encrypted_key": base64URLencode(libsodium.crypto_box(my_key, their_vk, cek, cek_iv))
+							^ this is crypto_box_easy(out, cek, rec_nonce, rec_pub, my_priv)
+	"header": {
+		"kid": "base58encode(recipient_verkey)",
+		"sender" : base64URLencode(libsodium.crypto_box_seal(their_vk, base58encode(sender_vk)),
+							^ this is crypto_box_seal(base58encode(my_pub), rec_pub)
+		"iv" : base64URLencode(cek_iv)
+	}
+}
+*/
+
+// buildRecipient encodes the necessary data for the recipient to decrypt the message
+// 	encrypting the CEK and sender pub key
+func (c *Crypter) buildRecipient(cek *[chacha20poly1305.KeySize]byte, recKey *publicEd25519) (*Recipient, error) {
+
+	var nonce [24]byte
+
+	_, err := rand.Reader.Read(nonce[:])
+	if err != nil {
+		return nil, err
+	}
+
+	senderSKCurve, err := secretEd25519toCurve25519(c.sender.priv)
+	if err != nil {
+		return nil, err
+	}
+
+	recPKCurve, err := publicEd25519toCurve25519(recKey)
+	if err != nil {
+		return nil, err
+	}
+
+	encCEK := box.Seal(nil, cek[:], &nonce, (*[32]byte)(recPKCurve), (*[32]byte)(senderSKCurve))
+
+	var encSender []byte
+	encSender, err = sodiumBoxSeal([]byte(base58.Encode(c.sender.pub[:])), recPKCurve)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Recipient{
+		EncryptedKey: base64.URLEncoding.EncodeToString(encCEK),
```

Comment:
Why not RawURLEncoding like the jwe package is doing?

Reply:
Encoding in raw base64 URL encoding, ie, without padding, is safe when the recipient knows exactly how long the data should be. I don't expect it would work when I'm dealing with a generic decoder function (in python) that isn't being told the length of the output data. I can test it.

Reply:
need a test with ref payload built in Python and test decryption  here

Reply:
These tests will come with #294 when the decrypter exists